### PR TITLE
[RFR] Add new entities

### PR DESCRIPTION
**Pull request by BaptistePDJ**

### CODE REVIEW

File: project/src/Controller/AdController.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace App\Controller;
+
+use App\Entity\Ad;
```
Comment (BaptistePDJ) [inline project/src/Controller/AdController.php:5]: Add setBinder

File: project/src/Controller/FrenchEchoppeController.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+use App\Entity\FrenchEchoppe;
+use App\Entity\Service;
+use App\Entity\SiteClientToServicesBinder;
+use App\Form\FrenchEchoppeType;
+use App\Repository\FrenchEchoppeRepository;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
```
Comment (BaptistePDJ) [inline project/src/Controller/FrenchEchoppeController.php]: merge those

File: project/src/Controller/MailController.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+use App\Entity\Mail;
+use App\Entity\Service;
+use App\Entity\SiteClientToServicesBinder;
+use App\Form\MailType;
+use App\Repository\MailRepository;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\{Request, Response};
+use Symfony\Component\Routing\Annotation\Route;
```
Comment (BaptistePDJ) [inline project/src/Controller/MailController.php]: merge those

File: project/src/Entity/SocialNetwork.php [hunk 4]
```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace App\Entity;
+
+use App\Repository\SocialNetworkRepository;
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
```
Comment (BaptistePDJ) [inline project/src/Entity/SocialNetwork.php]: merge

File: project/src/Form/SocialNetworkType.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+namespace App\Form;
+
+use App\Entity\Client;
+use App\Entity\SocialNetwork;
+use Symfony\Bridge\Doctrine\Form\Type\EntityType;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\Extension\Core\Type\NumberType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
```
Comment (BaptistePDJ) [inline project/src/Form/SocialNetworkType.php]: merge those

File: project/templates/click_and_collect/index.html.twig [hunk 6]
```diff
@@ -50,41 +52,40 @@
...
                                 <a href="{{ path('click_and_collect_show', {'id': click_and_collect.id}) }}" class="btn btn-secondary">Détails</a>
                             </td>
                         </tr>
-                    {% else %}
-                        <tr>
-                            <td colspan="13">Aucun Click & Collect</td>
-                        </tr>
                     {% endfor %}
                     </tbody>
                 </table>
```
Comment (BaptistePDJ) [inline project/templates/click_and_collect/index.html.twig:71]: else statement and endif

File: project/templates/french_echoppe/details.html.twig [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        <td>{{ currentFrenchEchoppe.renewalDate ? currentFrenchEchoppe.renewalDate|date('d-m-Y') : '' }}</td>
+    </tr>
+    <tr>
+        <th>Client</th>
+        <td>{{ currentFrenchEchoppe.client.name }}
+            <a href="{{ path('client_show', {'id': currentFrenchEchoppe.client.id}) }}" class="a-link btn icon-title-left" data-toggle="tooltip" title="Cliquez pour plus de détails"> <i class="fa fa-external-link" aria-hidden="true"></i></a>
+        </td>
+    </tr>
+    <tr>
+        <th>Server</th>
```
Comment (BaptistePDJ) [inline project/templates/french_echoppe/details.html.twig]: Serveur

Comment (BaptistePDJ) [reply project/templates/french_echoppe/details.html.twig]: Remove server

File: project/src/Controller/FrenchEchoppeController.php [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            ->setType(Service::FRENCH_ECHOPPE)
+            ->setServiceId($frenchEchoppe->getId());
+
+        $entityManager->persist($siteClientToServicesBinder);
+        $entityManager->flush();
+        $frenchEchoppe->setSiteClientToServicesBinderId($siteClientToServicesBinder->getId());
+
+        $entityManager->persist($frenchEchoppe);
+        $entityManager->flush();
+    }
```
Comment (julienmilazzo) [inline project/src/Controller/FrenchEchoppeController.php]: Use your util class no ?

File: project/src/Controller/MailController.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            ->setType(Service::MAIL)
+            ->setServiceId($mail->getId());
+
+        $entityManager->persist($siteClientToServicesBinder);
+        $entityManager->flush();
+        $mail->setSiteClientToServicesBinderId($siteClientToServicesBinder->getId());
+
+        $entityManager->persist($mail);
+        $entityManager->flush();
+    }
```
Comment (julienmilazzo) [inline project/src/Controller/MailController.php]: Same here ?

File: project/src/Controller/SocialNetworkController.php [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            ->setType(Service::SOCIAL_NETWORK)
+            ->setServiceId($socialNetwork->getId());
+
+        $entityManager->persist($siteClientToServicesBinder);
+        $entityManager->flush();
+        $socialNetwork->setSiteClientToServicesBinderId($siteClientToServicesBinder->getId());
+
+        $entityManager->persist($socialNetwork);
+        $entityManager->flush();
+    }
```
Comment (julienmilazzo) [inline project/src/Controller/SocialNetworkController.php]: same here ;)

File: project/src/Entity/SocialNetwork.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    private $client;
+
+    /**
+     * @ORM\OneToMany(targetEntity=Ad::class, mappedBy="socialNetwork", cascade={"persist"})
+     */
+    private $ad;
+
+    /**
+     *
+     */
```
Comment (julienmilazzo) [inline project/src/Entity/SocialNetwork.php]: You can remove this

File: project/src/Form/AdType.php [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace App\Form;
+
+use App\Entity\{Ad, Client, Site, SocialNetwork};
+use Symfony\Bridge\Doctrine\Form\Type\EntityType;
+use Symfony\Component\Form\Extension\Core\Type\DateType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class AdType extends ServiceType
```
Comment (julienmilazzo) [inline project/src/Form/AdType.php:11]: Add php doc for every function please

File: project/src/Repository/AdRepository.php [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    public function findOneBySomeField($value): ?GoogleAds
+    {
+        return $this->createQueryBuilder('g')
+            ->andWhere('g.exampleField = :val')
+            ->setParameter('val', $value)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+    }
+    */
```
Comment (julienmilazzo) [inline project/src/Repository/AdRepository.php]: You remoev those examples

File: project/src/Repository/FrenchEchoppeRepository.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    public function findOneBySomeField($value): ?FrenchEchoppe
+    {
+        return $this->createQueryBuilder('f')
+            ->andWhere('f.exampleField = :val')
+            ->setParameter('val', $value)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+    }
+    */
```
Comment (julienmilazzo) [inline project/src/Repository/FrenchEchoppeRepository.php]: same heere

File: project/src/Repository/MailRepository.php [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    public function findOneBySomeField($value): ?Mail
+    {
+        return $this->createQueryBuilder('m')
+            ->andWhere('m.exampleField = :val')
+            ->setParameter('val', $value)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+    }
+    */
```
Comment (julienmilazzo) [inline project/src/Repository/MailRepository.php]: Same

File: project/src/Repository/SocialNetworkRepository.php [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    public function findOneBySomeField($value): ?SocialNetwork
+    {
+        return $this->createQueryBuilder('s')
+            ->andWhere('s.exampleField = :val')
+            ->setParameter('val', $value)
+            ->getQuery()
+            ->getOneOrNullResult()
+        ;
+    }
+    */
```
Comment (julienmilazzo) [inline project/src/Repository/SocialNetworkRepository.php]: same

File: project/src/Controller/AdController.php [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+            ->setType(Service::AD)
+            ->setServiceId($ad->getId());
+
+        $entityManager->persist($siteClientToServicesBinder);
+        $entityManager->flush();
+        $ad->setSiteClientToServicesBinderId($siteClientToServicesBinder->getId());
+
+        $entityManager->persist($ad);
+        $entityManager->flush();
+    }
```
Comment (julienmilazzo) [inline project/src/Controller/AdController.php]: same

File: project/src/Controller/AdController.php [hunk 18]
```diff
@@ -2,9 +2,10 @@
 
 namespace App\Controller;
 
-use App\Entity\{Ad, Service, SiteClientToServicesBinder};
```
Comment (BaptistePDJ) [inline project/src/Controller/AdController.php]: Do this in all class

File: project/src/Util/Binder.php [hunk 19]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace App\Util;
+
+use App\Entity\Service;
+use App\Entity\SiteClientToServicesBinder;
```
Comment (BaptistePDJ) [inline project/src/Util/Binder.php]: merge those

File: project/src/Repository/AdRepository.php [hunk 20]
```diff
@@ -18,33 +18,4 @@ public function __construct(ManagerRegistry $registry)
...
-    public function findOneBySomeField($value): ?GoogleAds
-    {
-        return $this->createQueryBuilder('g')
-            ->andWhere('g.exampleField = :val')
-            ->setParameter('val', $value)
-            ->getQuery()
-            ->getOneOrNullResult()
-        ;
-    }
-    */
```
Comment (julienmilazzo) [inline project/src/Repository/AdRepository.php]: You can remove tbhis ;)

File: project/src/Repository/FrenchEchoppeRepository.php [hunk 21]
```diff
@@ -18,33 +18,4 @@ public function __construct(ManagerRegistry $registry)
...
-    /*
-    public function findOneBySomeField($value): ?FrenchEchoppe
-    {
-        return $this->createQueryBuilder('f')
-            ->andWhere('f.exampleField = :val')
-            ->setParameter('val', $value)
-            ->getQuery()
-            ->getOneOrNullResult()
-        ;
-    }
```
Comment (julienmilazzo) [inline project/src/Repository/FrenchEchoppeRepository.php]: same here

File: project/src/Repository/MailRepository.php [hunk 22]
```diff
@@ -18,33 +18,4 @@ public function __construct(ManagerRegistry $registry)
...
-    public function findOneBySomeField($value): ?Mail
-    {
-        return $this->createQueryBuilder('m')
-            ->andWhere('m.exampleField = :val')
-            ->setParameter('val', $value)
-            ->getQuery()
-            ->getOneOrNullResult()
-        ;
-    }
-    */
```
Comment (julienmilazzo) [inline project/src/Repository/MailRepository.php]: same

File: project/src/Repository/SocialNetworkRepository.php [hunk 23]
```diff
@@ -18,33 +18,4 @@ public function __construct(ManagerRegistry $registry)
...
-    public function findOneBySomeField($value): ?SocialNetwork
-    {
-        return $this->createQueryBuilder('s')
-            ->andWhere('s.exampleField = :val')
-            ->setParameter('val', $value)
-            ->getQuery()
-            ->getOneOrNullResult()
-        ;
-    }
-    */
```
Comment (julienmilazzo) [inline project/src/Repository/SocialNetworkRepository.php]: same

File: project/src/Util/Binder.php [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+     * @return void
+     */
+    static function set($service, EntityManagerInterface $entityManager)
+    {
+        $siteClientToServicesBinder = new SiteClientToServicesBinder();
+        $siteClientToServicesBinder
+            ->setClient($service->getClient())
+            ->setServiceId($service->getId());
+
+        if ('App\Entity\Ad' === get_class($service) || 'App\Entity\ClickAndCollect' === get_class($service) || 'App\Entity\DomainName' === get_class($service)) {
```
Comment (julienmilazzo) [inline project/src/Util/Binder.php]:
```
$classes = [Ad::class, ClickAndCollect::class, DomainName::class];
if (count(array_filter($classes, function ($class) use($service) { return !($service instanceof $class); }))) 
```

File: project/src/Util/Binder.php [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+        }
+
+        match (get_class($service)) {
+            'App\Entity\Ad' => $siteClientToServicesBinder->setType(Service::AD),
+            'App\Entity\SocialNetwork' => $siteClientToServicesBinder->setType(Service::SOCIAL_NETWORK),
+            'App\Entity\ClickAndCollect' => $siteClientToServicesBinder->setType(Service::CLICK_AND_COLLECT),
+            'App\Entity\FrenchEchoppe' => $siteClientToServicesBinder->setType(Service::FRENCH_ECHOPPE),
+            'App\Entity\DomainName' => $siteClientToServicesBinder->setType(Service::DOMAIN_NAME),
+            'App\Entity\Mail' => $siteClientToServicesBinder->setType(Service::MAIL),
+        };
```
Comment (julienmilazzo) [inline project/src/Util/Binder.php]:
```
$serviceRepositoryMapping = [
    Ad::class => Service::AD,
    SocialNetwork::class => Service::SOCIAL_NETWORK,
    ClickAndCollect::class => Service::CLICK_AND_COLLECT,
    FrenchEchoppe::class => Service::FRENCH_ECHOPPE,
    DomainName::class => Service::DOMAIN_NAME,
    DomainName::Mail => Service::MAIL,
];
$siteClientToServicesBinder->setType($serviceMapping[get_class($service)]),
```

**Review by julienmilazzo — APPROVED** [review]

**Pull request merged**